Source/missiles.cpp

```diff
@@ -4133,39 +4083,38 @@ void MI_Guardian(int i)
 		SetMissDir(i, 1);
 	}
 
+	Point position = missile[i].position.tile;
+
 	if ((missile[i]._mirange % 16) == 0) {
-		ex = 0;
-		for (j = 0; j < 23 && ex != -1; j++) {
-			for (k = 10; k >= 0 && ex != -1 && (vCrawlTable[j][k] != 0 || vCrawlTable[j][k + 1] != 0); k -= 2) {
-				if (sx1 == vCrawlTable[j][k] && sy1 == vCrawlTable[j][k + 1]) {
+		Displacement previous = { 0, 0 };
+
+		int ex = 0;
+		for (int j = 0; j < 23 && ex != -1; j++) {
+			for (int k = 10; k >= 0 && ex != -1; k -= 2) {
+				const Displacement offset { vCrawlTable[j][k], vCrawlTable[j][k + 1] };
+				if (offset == Displacement { 0, 0 }) {
+					break;
+				}
+				if (previous == offset) {
 					continue;
 				}
-				sx = missile[i].position.tile.x + vCrawlTable[j][k];
-				sy = missile[i].position.tile.y + vCrawlTable[j][k + 1];
-				ex = Sentfire(i, { sx, sy });
+				ex = Sentfire(i, { position.x + offset.deltaX, position.y + offset.deltaY });

```

Comment:
With how this has been refactored to use `Point`/`Displacement` types I think an element-wise (Hadamard) product function on Displacement could help explain what the tests are doing.

```cpp
/**
 * @brief Computes the element-wise product of two Displacements.
 *
 * @return { a.deltaX * b.deltaX, a.deltaY * b.deltaY }
 */
constexpr friend Displacement HadamardProduct(Displacement a, Displacement b) {
    return { a.deltaX * b.deltaX, a.deltaY * b.deltaY };
}

constexpr friend Displacement HadamardProduct(Displacement a, Direction dir) {
    return a.HadamardProduct(Displacement::fromDirection(dir));
}
```

```cpp
Sentfire(i, position + offset.HadamardProduct(DIR_S))
```

It's a very unusual operation though so there's likely a way to refactor this further to avoid needing it.

Not something I'd worry about now, I'd log an issue for this suggestion if I was more confident it's even useful.

Reply:
I think I will keep things as they are for now, but you are welcome to make this change.

Reply:
This is the first time I heard about Hadamard product. Interesting stuff.

I think it is a good idea to have abstractions like this to make the code simpler to reason about, so +1 from me.